### ISSUE TITLE
Feature request:  `"string" + XXX_strerror(...)` should be detected as an error

### ISSUE THREAD

**Issue by client9** [phpmd/phpmd#349]
Hello!

Here's an idea for a new rule, that I'm 100% positive will find real bugs. 

We had some networking code that was returning a  "0" for an error message.  It took while to track down.  The source the problem was the `+` instead of a `.`

``` php
$errmsg = "networking error: " + socket_strerror(...);
```

This type of bug is especially bad since it's often very difficult or impossible to create a test case for, so you find this bug in the worst possible way:  when you are having a real problem, and you get "0" for a log message :-(

There are a large number of `XXX_strerror()` functions that always return a string, not a number.
- http://php.net/manual/en/function.socket-strerror.php
- http://php.net/manual/en/function.posix-strerror.php
- http://php.net/manual/en/function.curl-strerror.php
- http://php.net/manual/en/function.pcntl-strerror.php
- http://php.net/manual/en/function.curl-multi-strerror.php
- http://php.net/manual/en/function.radius-strerror.php
- http://php.net/manual/en/function.mqseries-strerror.php

`error_get_last` is different but might also make sense to include.  http://php.net/manual/en/function.error-get-last.php

Doing `grep` in our code base found a mix of the correct use of `.` and incorrect use of `+`.  I have a hard time seeing why you would _ever_ use `+` with these functions, so it seems like it might be a good rule.  

I would contribute a PR but as you can probably tell, I'm not a regular PHP user!

Thanks for your time.

Regards,

nickg

**Comment by ravage84** [comment]
@client9 thanks for your idea.
I see your problem but creating a rule for covering such a narrow case would be an overkill.
Or better said not many will ever benefit from it.

The real problem here is that two strings are getting concatenated with the wrong operator.

My IDE, PhpStorm, for example, shows me this clearly.

![string addition](https://cloud.githubusercontent.com/assets/625761/12987014/739cc90a-d0fa-11e5-87ae-116673ec17c6.png)

Also, I can use PhpStorm's built-in code inspection over my whole project, which would uncover me such mistakes.

I don't say PHPMD is the wrong tool for that or that we shouldn't create a rule for the problem at hand, but there are existing tools, which are in the better position as PHPMD is.
Because an IDE does a full index of your code and has also an internal index of the PHP function signatures.

**Comment by client9** [comment]
Hi guys,

thanks for your answer.

Ideally what I'm looking for is "PhpStorm analysis", but done via command line so it can go into a automated CI system.

That's probably NOT the goal of PHPMD, so it's ok to close this if you wish.

thanks for listening, and thanks for educating me!

regards,

n

**Comment by AJenbo** [comment]
Probably better to just use PHPStan for this.
